libkbfs/block_retrieval_queue.go

```diff
@@ -74,18 +78,38 @@ func newBlockRetrievalQueue(numWorkers int) *blockRetrievalQueue {
 // notifyWorker notifies workers that there is a new request for processing.
 func (brq *blockRetrievalQueue) notifyWorker() {
 	go func() {
-		// Get the next queued worker
-		ch := <-brq.workerQueue
-		// Prevent interference with the heap while we're retrieving from it
-		brq.mtx.Lock()
-		defer brq.mtx.Unlock()
-		// Pop from the heap
-		ch <- heap.Pop(brq.heap).(*blockRetrieval)
+		select {
+		case <-brq.doneCh:
+			// Prevent interference with the heap while we're retrieving from it
+			brq.mtx.Lock()
+			defer brq.mtx.Unlock()
+			if brq.heap.Len() > 0 {
```

Comment:
I guess you'll have the same locking issue whether it's an `if` or a `for`, right?  You could make a helper `popIfAnyAvailable` function that takes the lock and only pops if len > 0.  (It would be nice if the heap's pop could work on an empty heap without panicking, but I guess it can't.)

I forgot there was exactly one of these goroutines for every request.  So if you want to keep it one-to-one, maybe add a comment here explaining that.